core/kernel/shared/sprintf.rb

```diff
@@ -0,0 +1,945 @@
+          format("%g", 1234.1234567).should == "1234.12"
+        end
+      end
+
+      it "displays Float::INFINITY as Inf" do
+        format("%g", Float::INFINITY).should == "Inf"
+        format("%g", -Float::INFINITY).should == "-Inf"
+      end
+
+      it "displays Float::NAN as NaN" do
+        format("%g", Float::NAN).should == "NaN"
+        format("%g", -Float::NAN).should == "NaN"
+      end
+    end
+
+    describe "G" do
+      context "the exponenta is less than -4" do
+        it "converts a floating point number using exponential form and use an uppercase E" do
+          format("%G", 0.0000123456).should == "1.23456E-05"
+          format("%G", -0.0000123456).should == "-1.23456E-05"
+
+          format("%G", 0.000000000123456).should == "1.23456E-10"
+          format("%G", -0.000000000123456).should == "-1.23456E-10"
+        end
+      end
+
+      context "the exponenta is greater than or equal to the precision (6 by default)" do
+        it "converts a floating point number using exponential form" do
+          format("%G", 1234567).should == "1.23457E+06"
+          format("%G", 1234567890123).should == "1.23457E+12"
+          format("%G", -1234567).should == "-1.23457E+06"
+        end
+      end
+
+      context "otherwise" do
+        it "converts a floating point number in dd.dddd form" do
+          format("%G", 0.0001).should == "0.0001"
+          format("%G", -0.0001).should == "-0.0001"
+          format("%G", 123456).should == "123456"
+          format("%G", -123456).should == "-123456"
+        end
+
+        it "cuts excessive digits in fractional part" do
+          format("%G", 12.12341111).should == "12.1234"
+          format("%G", -12.12341111).should == "-12.1234"
+        end
+
+        it "rounds the last significant digit to the closest one in fractional part" do
+          format("%G", 1.555555555).should == "1.55556"
+          format("%G", -1.555555555).should == "-1.55556"
+          format("%G", 1.444444444).should == "1.44444"
+        end
+
+        # strange behavior
+        it "cuts fraction part to have only 6 digits at all" do
+          format("%G", 1.1234567).should == "1.12346"
+          format("%G", 12.1234567).should == "12.1235"
+          format("%G", 123.1234567).should == "123.123"
+          format("%G", 1234.1234567).should == "1234.12"
+        end
+      end
+
+      it "displays Float::INFINITY as Inf" do
+        format("%G", Float::INFINITY).should == "Inf"
+        format("%G", -Float::INFINITY).should == "-Inf"
+      end
+
+      it "displays Float::NAN as NaN" do
+        format("%G", Float::NAN).should == "NaN"
+        format("%G", -Float::NAN).should == "NaN"
+      end
+    end
+
+    describe "a" do
+      # strange behavior
+      # from docs - Convert floating point argument as [-]0xh.hhhhp[+-]dd
+      # * p+7 => p+07 - two digits for exponenta
+      # * fraction part should be 4 digits long by default
```

Comment:
The behavior is probably correct and the documentation wrong. I think this just follows C's sprintf.
The man page is more verbose but not very clear either.
I think notations like `[-]0xh.hhhhp[+-]dd` don't mean exactly 4 hexdigits after the dot, more like the general form.
```
       a, A   (C99; not in SUSv2, but added in SUSv3) For a conversion, the double  argument  is  con‐
              verted  to  hexadecimal  notation (using the letters abcdef) in the style [-]0xh.hhhhp±;
              for A conversion the prefix 0X, the letters ABCDEF, and  the  exponent  separator  P  is
              used.  There is one hexadecimal digit before the decimal point, and the number of digits
              after it is equal to the precision.  The default precision suffices for an exact  repre‐
              sentation of the value if an exact representation in base 2 exists and otherwise is suf‐
              ficiently large to distinguish values of type double.   The  digit  before  the  decimal
              point  is  unspecified  for nonnormalized numbers, and nonzero but otherwise unspecified
              for normalized numbers.
```

Reply:
Agreed

Reply:
@eregon @marcandre 

documentation is confusing me because usually format like `dd.dddd` means exactly no more than 4 digits after point in other format descriptions.